rust/tools/authorization-logic/transparent_release_prototype/transparent_release.auth_logic

```diff
@@ -0,0 +1,276 @@
+//  the following.
+//
+// The arguments in this example are:
+// - provenance file: "provenance_file.toml"
+// - principal name for application: "OakFunctionsLoader"
+
+"ProvenanceFileBuilder" says {
+    "Binary_OakFunctionsLoader"
+        hasProvenance("ProvenanceFile_OakFunctionsLoader").
+}
+
+//-----------------------------------------------------------------------------
+// Hash Check Wrapper (Checks hashes on files)
+//-----------------------------------------------------------------------------
+// We need a wrapper (another standalone program) that outputs a measurement
+// (sha256) of various files as authorization logic (provenance file, binary).
+// This wrapper will compute the sha256 directly (for example, using the CLI program
+// "sha256sum", or by calling 
+// [the code in the transparent release repository that checks
+// hashes](https://github.com/project-oak/transparent-release/blob/2c8f8292cab31dbacbf5f4352b51025599ae78c3/common/common.go#L316)
+// ).
+//
+// The hash check wrapper for the binary can run on the "output_path" from 
+// transparent-release/common/BuildConfig, which is where the binary is stored 
+// after it is built.
+//
+// (You could imagine extending transparent-release/common::Build() to just output 
+// the hash instead, but then there would be two principals trusted for 
+// producing hashes instead of just one, so the trust assumptions would be less 
+// clear.)
+
+// The arguments to this program are:
+// - the name of the file to be measured (hash checked)
+// - the name of the principal for the file being measured
+
+// We will invoke this wrapper twice: once for the provenance file and once
+// for the binary generated by executing the steps in the provenance file
+
+// For the provenance file, the inputs are:
+//    file name --> "provenance_file.toml"
+//    principal name --> "ProvenanceFile_OakFunctionsLoader"
+// The output looks like:
+"Sha256Wrapper" says {
+    measured_hash("ProvenanceFile_OakFunctionsLoader",
+      "6de656f11599c11548519af6b7385504ed845a4ff6600745cb3d9e174b8956af").
+}
+
+// For the binary, the inputs are:
+//    file name --> "oak_functions_loader.exe" (I don't have an example binary)
+//    principal name --> "Binary_OakFunctionsLoader"
+"Sha256Wrapper" says {
+    measured_hash("Binary_OakFunctionsLoader",
+      "cb255dbbea86b128379995ff7dc805fd026e2d21c997a8df688ba9e5cb98e1c1").
+}
+
+//-----------------------------------------------------------------------------
+// Unix Time Wrapper
+//-----------------------------------------------------------------------------
+// We need a wrapper that produces the real time in unix epoch seconds. A 
+// command for doing this on UNIX systems is `date %s`. Assuming the time
+// that this command is run is 2021-10-21, 19:14:58 the output from running
+// this command is 1634843698.
+//
+// This wrapper is also a standalone binary. It takes no inputs and outputs
+// authorization logic code that looks like this:
```

Comment:
What you are suggesting is not what I originally intended. I recall you suggested this when we spoke and I though it made sense at the time. Now I don't think this suggestion changes anything about the TCB, but it adds extra parts and complexity. So now I think it is better not to do this and just to have the principal names emitted directly. 

If wrappers just emit the inner statements but not the `"X says"` part then we need another "outer wrapper" or something similar that attaches `"X says"` to the inner part. In this case we still have to trust the "outer wrapper" to not do this impersonation. So I don't think this changes the trust assumptions.

Maybe we can get what you want by just tweaking the interface of the calls that the top level makes so that the full principal name is given. For example, the interface for the EndorsementFile wrapper is:

```
EndorsementFileWrapper(EndorsementFilePath, AppName)
```

and the name of the principal to generate is `StrCat("EndorsementFile_", AppName)`

but maybe you are happier if instead just the whole name of the principal for the endorsement file is passed and it's:

```
EndorsementFileWrapper(EndorsementFilePath, PrincipalName)
```

(I think that's about the same as what you want, but it's fewer parts).

Reply:
I do think there is a difference though. In the current design, any wrapper (top-level or inner) can impersonate any other wrapper. e.g. the sha256wrapper may emit 

```
"UnixEpochTime" says {
    RealTimeIs("1634843698").
}
```

and it would considered valid. If instead we limit each inner wrapper to emit only the things that it actually wants to say, the top-level wrapper can be trusted to accurately reflect who says what in the final output. Which makes sense, because the top-level wrapper is already trusted to call the appropriate inner wrappers, so it should be its responsibility to keep track of the speakers.

I think this would limit the amount of trust that needs to be granted to each wrapper, right?

Reply:
Ok so I think what you are saying is like:


### Old Design
We change from:

`EndorsementFileWrapper(EndorsementFilePath, AppName)` which produces "EndorsementFile_#{AppName}" says { Blah(). }`

to:
### New Design
`EndorsementFileWrapper(EndorsementFilePath)` which produces just `Blah().`

and then we just add another function like:

`WrapWithPrincipal(AuthLogicBeliefs, PrincipalName)` which gives `"PrincipalName" says { AuthLogicBeliefs }`

and the top level calls:

`WrapWithPrincipal(EndorsementFileWrapper(EndorsementFilePath), "EndorsementFile_AppName")`

If so OK yes this is good let's do this.

Reply:
Yes exactly, thanks for the accurate summary! I think this way each wrapper is much less trusted to get things right.